overlord/ifacestate/hotplug.go

```diff
@@ -86,14 +106,210 @@ func defaultDeviceKey(devinfo *hotplug.HotplugDeviceInfo, keyVersion int) (strin
 
 // hotplugDeviceAdded gets called when a device is added to the system.
 func (m *InterfaceManager) hotplugDeviceAdded(devinfo *hotplug.HotplugDeviceInfo) {
+	st := m.state
+	st.Lock()
+	defer st.Unlock()
+
+	if _, err := snapstate.CoreInfo(st); err != nil {
+		logger.Noticef("core snap not available, hotplug events ignored")
+		return
+	}
+
+	hotplugIfaces := m.repo.AllHotplugInterfaces()
+	defaultKey, err := defaultDeviceKey(devinfo, deviceKeyVersion)
+	if err != nil {
+		logger.Noticef("cannot compute default hotplug key for device with path %s: %s", devinfo.DevicePath(), err.Error())
+	}
+
+	hotplugFeature, err := m.hotplugEnabled()
+	if err != nil {
+		logger.Noticef("internal error: cannot get hotplug feature flag: %s", err.Error())
+		return
+	}
+
+	gadget, err := snapstate.GadgetInfo(st)
+	if err != nil && err != state.ErrNoState {
+		logger.Noticef("internal error: cannot get gadget information: %s", err)
+	}
+
+	gadgetSlotsByInterface := make(map[string][]*snap.SlotInfo)
+	if gadget != nil {
+		ifcs := make(map[string]bool)
+		for _, iface := range hotplugIfaces {
+			ifcs[iface.Name()] = true
+		}
+		for _, gadgetSlot := range gadget.Slots {
+			if _, ok := ifcs[gadgetSlot.Interface]; ok {
+				gadgetSlotsByInterface[gadgetSlot.Interface] = append(gadgetSlotsByInterface[gadgetSlot.Interface], gadgetSlot)
+			}
+		}
+	}
+
+InterfacesLoop:
+	// iterate over all hotplug interfaces
+	for _, iface := range hotplugIfaces {
+		hotplugHandler := iface.(hotplug.Definer)
+
+		// determine device key for the interface; note that interface might provide own device keys.
+		key, err := deviceKey(defaultKey, devinfo, iface)
+		if err != nil {
+			logger.Noticef("cannot compute hotplug key for device with path %s: %s", devinfo.DevicePath(), err.Error())
+			continue
+		}
+
+		// ignore device that is already handled by a gadget slot
+		if gadgetSlots, ok := gadgetSlotsByInterface[iface.Name()]; ok {
+			for _, gslot := range gadgetSlots {
+				if pred, ok := iface.(hotplug.HandledByGadgetPredicate); ok {
+					if pred.HandledByGadget(devinfo, gslot) {
+						logger.Debugf("ignoring device with hotplug key %q, interface %s (handled by gadget slot %s)", key, iface.Name(), gslot.Name)
+						continue InterfacesLoop
+					}
+				}
+			}
+		}
+
+		proposedSlot, err := hotplugHandler.HotplugDeviceDetected(devinfo)
+		if err != nil {
+			logger.Noticef("cannot process hotplug event by the rule of interface %q: %s", iface.Name(), err)
+			continue
+		}
+		if proposedSlot == nil {
+			continue
+		}
+		if key == "" {
```

Comment:
why check this so far from where we got it?

Reply:
Right, no good reason, I think this code just drifted away as I changed this loop a few times. Will fix. Thanks

Reply:
Actually, now as I'm looking at spread test in the other branch I remember why I did it that way: it's useful to know when a key couldn't be computed and log it, but only if the interface actually wants to create a slot. Doing this early creates a lot of log noise. I'll move this check back, with a comment.

Reply:
Think more about it, the key can simply be computed after proposedSlot and close to the key == "" check. Done.